### PR TITLE
"Training to inference" tutorial

### PR DESCRIPTION
**Pull request by szalpal**
This PR introduces a `training -> inference` tutorial. The use-case is based on Efficientnet model, presented both in DeepLearningExamples and DALI repository. All assumptions and explanations are provided inside.

### CODE REVIEW

File: docs/tutorials/training_to_inference.md [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+This tutorial presents the simple way of going from training to inference. We are focusing on a particular scenario: 
+how to deploy a model, when it has been trained using DALI as a preprocessing tool. An example that constitutes this 
+tutorial is the [EfficientNet](https://github.com/NVIDIA/DALI/tree/main/docs/examples/use_cases/pytorch/efficientnet) 
+from DALI repository. This tutorial is split in two parts - [Theory](#Theory) and [Practice](#Practice). Lastly, while we cover only the 
+deployment of DALI preprocessing graph, we are using handful of other NVIDIA tools useful for setting up the inference. 
+Documentation and links to those tools will be provided, but the specifics of these tools won't be explained - should 
+you like to know more details about them, please refer to the documentation.
+
+## Assumptions
```
Comment (banasraf) [inline docs/tutorials/training_to_inference.md:11]: I feel like this section could be a first subsection of the Practice part. The theory is more general and those assumptions refer to a specific use case

File: docs/tutorials/training_to_inference.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                       antialias=False)
+
+    output = fn.crop_mirror_normalize(images, dtype=types.FLOAT, output_layout=output_layout,
+                                      crop=(image_crop, image_crop),
+                                      mean=[0.485 * 255, 0.456 * 255, 0.406 * 255],
+                                      std=[0.229 * 255, 0.224 * 255, 0.225 * 255])
+    return output, label
+```
+
+We're adjusting it by swapping the Reader with ExternalSource and adding a `name` parameter to it. Since in our example
```
Comment (banasraf) [inline docs/tutorials/training_to_inference.md:73]:
```suggestion
We're adjusting it by swapping the reader with external_source and adding a `name` parameter to it. Since in our example 
```

just a suggestion. PascalCase names seem kinda outdated

Comment (banasraf) [reply docs/tutorials/training_to_inference.md:73]: or even just `external source`

File: docs/tutorials/training_to_inference.md (hunk 1 above)
Comment (szalpal) [reply docs/tutorials/training_to_inference.md:11]: Good point. My reasoning was that these assumptions also refer to the `theory` part. Especially: `@pipeline_def` requirement on the pipeline definition, using Triton for the inference orchestration and putting preprocessing on the server side. Do you feel that the Theory part without emphasising these would be understandable?

**Review by banasraf — APPROVED** [review]

File: docs/tutorials/training_to_inference.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    1. Insert the DALI pipeline definition (prepared in the **Step 1**) into the `dali.py` file in the version directory.
+    1. Create a [configuration file](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/model_configuration.html) (unnecessary, if you are using Model Autoconfiguration for DALI Backend). Make sure that the names of the inputs in the model configuration match the names of the input operators assigned in **Step 1**.
+    1. Combine the Deep Learning and DALI models using [model ensemble](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/architecture.html#ensemble-models) or [BLS script](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/index.html?highlight=business%20logic%20scripting).
+1. Run the Triton server and send some requests to it.
+
+Enough with the theory. Let's go to practice.
+
+## Practice
+
+### ➔ 1: Adjusting DALI Pipeline
```
Comment (awolant) [inline docs/tutorials/training_to_inference.md:49]: An idea that came to my mind while reading this section: maybe it would be useful and not that hard to do to create a sort of linter tool/script to analyze DALI pipeline and mark parts that maybe should be adjusted for the inference? Like it goes through the graph, sees some random op and prints a warning: "You have a RandomCrop defined at line 54" and so on. This could codify these best practices in a tool.

But as I said, just an idea, not really a comment to this PR :)

**Review by awolant — DISMISSED** [review]

File: docs/tutorials/training_to_inference.md [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+Documentation and links to those tools will be provided, but the specifics of these tools won't be explained - should 
+you like to know more details about them, please refer to the documentation.
+
+## Assumptions
+
+Since we are tackling a specific use-case in this tutorial, we will set up several assumptions:
+1. We are deploying an EfficientNet model, trained with PyTorch. The checkpoint is saved using `torch.save` and DALI 
+has been used as a preprocessing tool.
+2. The input data to this model reflects the input data used during training. To be specific - we will infer on encoded 
+JPEG files.
```
Comment (JanuszL) [inline docs/tutorials/training_to_inference.md:17]:
```suggestion
images.
```

File: docs/tutorials/training_to_inference.md [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+3. The training script contains DALI model, implemented using DALI's `@pipeline_def` decorator. Should you have the 
+DALI pipeline implemented using other approach - please convert to `@pipeline_def` first.
+4. Triton Inference Server will be used to orchestrate the inference.
+5. The entire preprocessing part of the inference will happen on the server side, using the GPU.
```
Comment (JanuszL) [inline docs/tutorials/training_to_inference.md:21]: I wonder if we should mention that the CPU processing is also supported. Maybe:
```suggestion
5. The entire preprocessing part of the inference will happen on the server side (using the GPU to speed things up).
```

File: docs/tutorials/training_to_inference.md [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+## Theory
+
+Unfortunately, it is not possible to introduce complete and precise algorithm for setting up the inference 
+(if it would, we'd just put together a script for it). However, the following is a highly useful "How-To" guide:
+1. Adjust a DALI Pipeline for the inference:
+    1. Make sure that the DALI pipeline you’ll be working on is the pipeline that contains the operations for the inference. Usually the **training** preprocessing and **inference** preprocessing differs - the former one contains some random operations that are used to augment the dataset. Most often the inference pipeline will match the **validation** pipeline (and not the training one).
```
Comment (JanuszL) [inline docs/tutorials/training_to_inference.md:28]: Can you link each of the theory part to the relevant `Practice` parts?

File: docs/tutorials/training_to_inference.md [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+                                      mean=[0.485 * 255, 0.456 * 255, 0.406 * 255],
+                                      std=[0.229 * 255, 0.224 * 255, 0.225 * 255])
+    return output, label
+```
+
+We're adjusting it by swapping the Reader with ExternalSource and adding a `name` parameter to it. Since in our example 
+we will explicitly create a `config.pbtxt` file for DALI model, we're not adding `ndim` and `dtype` arguments to 
+`fn.external_source`. We are, however, using [Autoserialization](https://github.com/triton-inference-server/dali_backend#autoserialization), 
+hence the `@autoserialize` decorator on top. Lastly, we are providing default values for arguments of the `inference_pipe`
+function.
```
Comment (JanuszL) [inline docs/tutorials/training_to_inference.md:77]: I would also mention that the labels are no longer a part of the pipeline output.

File: docs/tutorials/training_to_inference.md [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+for any details about running the server and leveraging Triton Client module to send requests to the server.
+
+### Triton Client hints
+
+As mentioned earlier, you can find the examples of Triton client implementation inside [DALI Backend repository](https://github.com/triton-inference-server/dali_backend/tree/main/qa). 
+The purpose of this tutorial is not to fully explain how to create a Triton client module. However, the following are 
+a couple of suggestions for creating a Triton client specifically for DALI model.
+
+#### Loading data as binary buffer
+NVIDIA hardware offers image and video decoding acceleration using [Hardware JPEG Decoder](https://developer.nvidia.com/blog/leveraging-hardware-jpeg-decoder-and-nvjpeg-on-a100/) and [NVDEC cores](https://developer.nvidia.com/video-codec-sdk).
```
Comment (JanuszL) [inline docs/tutorials/training_to_inference.md:175]: I would make this broader. We have also nvJPEG2k and nvTIFF, which uses just SMs. So how about:
```suggestion
NVIDIA offers GPU accelerated image and video decoding using various software (i.e. [nvJPEG2k](https://docs.nvidia.com/cuda/nvjpeg2000/userguide.html) or [nvTIFF](https://docs.nvidia.com/cuda/nvtiff/)) and hardware (like [JPEG](https://developer.nvidia.com/blog/leveraging-hardware-jpeg-decoder-and-nvjpeg-on-a100/) and [NVDEC](https://developer.nvidia.com/video-codec-sdk)) solutions.```

File: docs/tutorials/training_to_inference.md [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+```python
+def load_image(image_path):
+    return np.fromfile(image_path, dtype=np.uint8)
+```
+
+#### Batching data in the request
+
+With Triton and DALI Backend you can leverage batching in two ways. Firstly, you can turn on [Dynamic Batching](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/model_configuration.html#dynamic-batcher). 
+With this feature, you send the samples one-by-one to the server and the server automatically puts together a batch of data for you.
```
Comment (JanuszL) [inline docs/tutorials/training_to_inference.md:188]: `one-by-one` or any lenght?

File: docs/tutorials/training_to_inference.md (hunk 1 above)
Comment (banasraf) [reply docs/tutorials/training_to_inference.md:11]: Hmm. For me the theory part is understandable, but it works both way, so maybe keep according to your intuition.

File: docs/tutorials/training_to_inference.md (hunk 6 above)
Comment (szalpal) [reply docs/tutorials/training_to_inference.md:28]: Done

File: docs/tutorials/training_to_inference.md (hunk 2 above)
Comment (szalpal) [reply docs/tutorials/training_to_inference.md:73]: Done

File: docs/tutorials/training_to_inference.md (hunk 4 above)
Comment (szalpal) [reply docs/tutorials/training_to_inference.md:17]: Done

File: docs/tutorials/training_to_inference.md (hunk 5 above)
Comment (szalpal) [reply docs/tutorials/training_to_inference.md:21]: Done

File: docs/tutorials/training_to_inference.md (hunk 7 above)
Comment (szalpal) [reply docs/tutorials/training_to_inference.md:77]: Done

File: docs/tutorials/training_to_inference.md (hunk 9 above)
Comment (szalpal) [reply docs/tutorials/training_to_inference.md:188]: In theory any length, but the most common usage is that the samples are sent one by one. Removed the `one-by-one`.

**Review by awolant — APPROVED** [review]

**Review by banasraf — APPROVED** [review]
